### ISSUE TITLE
The library size is very large

### ISSUE THREAD

**Issue by denesbartha** [zezo357/pytorch_lite#57]
I would like to use the library `pytorch_lite` for object detection using a YOLO8 model.

However I saw that the size of the APK file is too large, and then I found that the file `libpytorch_jni.so` is ~74MB.
It does not matter whether I specify the version explicitly `pytorch_lite: ^4.2.1` or I provide the `path: ../` to the `pytorch_lite` package, the result is roughly the same.

Is there a way to shrink the lib size somehow by removing some unnecessary dependencies?

**Comment by zezo357** [comment]
Sadly I couldn't find a way to do so, prs are always welcomed in this matter

**Comment by noob8boi** [comment]
> I would like to use the library `pytorch_lite` for object detection using a YOLO8 model.
> 
> However I saw that the size of the APK file is too large, and then I found that the file `libpytorch_jni.so` is ~74MB. It does not matter whether I specify the version explicitly `pytorch_lite: ^4.2.1` or I provide the `path: ../` to the `pytorch_lite` package, the result is roughly the same.
> 
> Is there a way to shrink the lib size somehow by removing some unnecessary dependencies?

I was facing the same issue and after some research found out that yolo v8 model can be exported with --optimize flag which will call the `_save_for_lite_interpreter` function and now the torchscript model will work with lite dependencies. After that I uncommented the dependencies inside the build.gradle file :- 

```
implementation 'org.pytorch:pytorch_android_lite:2.1.0'
implementation 'org.pytorch:pytorch_android_torchvision_lite:2.1.0'
```
and comment these :- 
```
implementation 'org.pytorch:pytorch_android:2.1.0'
implementation 'org.pytorch:pytorch_android_torchvision:2.1.0'
```

With that you also have to use `LiteModuleLoader.load` instead of `Module.load` for loading your model in `PytorchLitePlugin.java` file & `import org.pytorch.LiteModuleLoader` in the same file.

**Comment by zezo357** [comment]
When I tried that the size didn't change much in my case , can you let me know the difference in size between the normal one and lite one

And if you don't mind can you open a pr?

**Comment by zezo357** [comment]
> > When I tried that the size didn't change much in my case , can you let me know the difference in size between the normal one and lite one
> > 
> > And if you don't mind can you open a pr?
> 
> earlier `libpytorch_jni.so` was around 70MB but now 18.7MB.. was it ~70MB for you before & after the change?
> ![image](https://github.com/zezo357/pytorch_lite/assets/91360152/20160c11-0346-4fbf-b1b3-7c00cc6de72d)
> 
> Regarding the PR.. should I make the code changes as mentioned above?
> 

Yes for the pr same changes you mentioned, and I meant the size of the apk using normal and _lite

**Comment by noob8boi** [comment]
> > > When I tried that the size didn't change much in my case , can you let me know the difference in size between the normal one and lite one
> > > And if you don't mind can you open a pr?
> > 
> > 
> > earlier `libpytorch_jni.so` was around 70MB but now 18.7MB.. was it ~70MB for you before & after the change?
> > ![image](https://private-user-images.githubusercontent.com/91360152/293186543-20160c11-0346-4fbf-b1b3-7c00cc6de72d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM3NjMyMTEsIm5iZiI6MTcwMzc2MjkxMSwicGF0aCI6Ii85MTM2MDE1Mi8yOTMxODY1NDMtMjAxNjBjMTEtMDM0Ni00ZmJmLWIxYjMtN2MwMGNjNmRlNzJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI4VDExMjgzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiMzYyM2U1NDI0ODRkMmY5ODMwNmQ2ZjkwN2MzOGVkMWI5Zjk0YjI2M2Y4ZjJlOGNjOTBiMjk5MGY3YWFjOTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8Q2V7UfA7VXHI-HrMOqtXUElFdQPFb46nsmX3T-V3IM)
> > Regarding the PR.. should I make the code changes as mentioned above?
> 
> Yes for the pr same changes you mentioned, and I meant the size of the apk using normal and _lite

![image](https://github.com/zezo357/pytorch_lite/assets/91360152/a813faba-84c7-40a6-a3b7-028dcff4aa43)

i recorded this yesterday but for some reason i am unable to reproduce the ~330MB build again even after deleting `pubspec.lock` , clearing all cache etc. Now I keep getting 130~ MB builds with/without the above changes. Maybe the code change isn't making much of a difference after all 🤔

**Comment by noob8boi** [comment]
Okay so `flutter build apk --target-platform android-arm64 --analyze-size` outputs two APKs and 20MB is saved with the lite dependency. Still not sure why I was getting 330MB yesterday. Anyway gonna create a PR now 👍
![image](https://github.com/zezo357/pytorch_lite/assets/91360152/b7eeda5f-c06c-4fc8-b7d7-04376fa71c03)
![image](https://github.com/zezo357/pytorch_lite/assets/91360152/f20f54c0-d64b-4331-8708-bb6c44197ba6)

**Comment by noob8boi** [comment]
https://github.com/zezo357/pytorch_lite/pull/68

**Comment by zezo357** [comment]
Thank you very much, will run workflows and test if everything good

**Comment by zezo357** [comment]
as i mentioned the difference is not much 5mb
you pr  https://github.com/zezo357/pytorch_lite/actions/runs/7348278520/job/20007199904
<img width="1198" alt="image" src="https://github.com/zezo357/pytorch_lite/assets/25157308/cd4b9d67-5f7f-4287-aeb1-7568f663332d">

before it 
https://github.com/zezo357/pytorch_lite/actions/runs/7185854080/job/19569974871
<img width="1168" alt="image" src="https://github.com/zezo357/pytorch_lite/assets/25157308/fa2f1420-9ec0-4598-b4fa-eaf455dbd810">


the problem is i dont know if the reduction in size hurts the models options that can be ran? and the reduction is roughly 5mb i dont know if it worth taking the risk

confirm with me if you any other opinion i think it can be published as a work in progress (beta) while we convert to pytorch lite and latest ios version if its stable i think it has some problems https://github.com/zezo357/pytorch_lite/issues/63

**Comment by noob8boi** [comment]
> as i mentioned the difference is not much 5mb you pr https://github.com/zezo357/pytorch_lite/actions/runs/7348278520/job/20007199904 <img alt="image" width="1198" src="https://private-user-images.githubusercontent.com/25157308/293255918-cd4b9d67-5f7f-4287-aeb1-7568f663332d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM3ODgxOTQsIm5iZiI6MTcwMzc4Nzg5NCwicGF0aCI6Ii8yNTE1NzMwOC8yOTMyNTU5MTgtY2Q0YjlkNjctNWY3Zi00Mjg3LWFlYjEtNzU2OGY2NjMzMzJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI4VDE4MjQ1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyMjg5YmFjZDQ4MjE2MTRlYjVlNTM2NWU1ZTI5ZTcwY2RjNjJiZGMxNjFiNTMzZmM5NzJhM2M4MWNiMTJhOGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5xK98X_imRPnapgpvyA_2mXa8NJyRStuNjwl9LXg7Sk">
> 
> before it https://github.com/zezo357/pytorch_lite/actions/runs/7185854080/job/19569974871 <img alt="image" width="1168" src="https://private-user-images.githubusercontent.com/25157308/293256050-fa2f1420-9ec0-4598-b4fa-eaf455dbd810.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM3ODgxOTQsIm5iZiI6MTcwMzc4Nzg5NCwicGF0aCI6Ii8yNTE1NzMwOC8yOTMyNTYwNTAtZmEyZjE0MjAtOWVjMC00NTk4LWI0ZmEtZWFmNDU1ZGJkODEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI4VDE4MjQ1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkwMzljZmZlNDYyYTZiNWRhMWI5NTA2ZGE0ODNjY2NmYTUwMmIyZjZhMWVkZWJkZTk1MDMxYjc4MTk4YjRiNmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.18MRqbYe_qHRFQZFSYfny5Bh56BIHQph15wJPU_AeJ0">
> 
> the problem is i dont know if the reduction in size hurts the models options that can be ran? and the reduction is roughly 5mb i dont know if it worth taking the risk
> 
> confirm with me if you any other opinion i think it can be published as a work in progress (beta) while we convert to pytorch lite and latest ios version if its stable i think it has some problems #63

I understand.. actually i posted the comment because i thought this code change will help in reducing `libpytorch_jni.so` file from 70MB to less than 15MB and solve OP's problem. But I have no clue what was the actual cause of it.

I guess the issue can remain open until further optimizations are found then we can consider creating another PR

**Comment by noob8boi** [comment]
if you don't mind can you build output apk file with --analyze-size flag. I want to see the space taken by assets and the `libpytorch_jni_lite.so`

**Comment by zezo357** [comment]
> if you don't mind can you build output apk file with --analyze-size flag. I want to see the space taken by assets and the `libpytorch_jni_lite.so`

Sadly I can't do it now, maybe tomorrow and I will let you know

**Comment by zezo357** [comment]
added it in master and in your pr

**Comment by zezo357** [comment]
pr:  https://github.com/zezo357/pytorch_lite/actions/runs/7351145699/job/20014024960
master:  https://github.com/zezo357/pytorch_lite/actions/runs/7351120618/job/20013953038

**Comment by noob8boi** [comment]
> ▒▒▒▒▒▒▒

As you can see 83MB is taken by the flutter assets which I believe is from the model. So I don't think the package can perform any further optimizations over there. And `armabi-v7a` file is only around 25MB.. so the `libpytorch_jni_lite.so` is definitely not 70MB .. more like 15MB or so. The analysis json file should be able give in depth view with the help of dart dev tools 🤔.. I hope I'm not drawing incorrect conclusions.

**Comment by zezo357** [comment]
> > ▒▒▒▒▒▒▒
> 
> As you can see 83MB is taken by the flutter assets which I believe is from the model. So I don't think the package can perform any further optimizations over there. And `armabi-v7a` file is only around 25MB.. so the `libpytorch_jni_lite.so` is definitely not 70MB .. more like 15MB or so. The analysis json file should be able give in depth view with the help of dart dev tools 🤔.. I hope I'm not drawing incorrect conclusions. Basically I don't think there can be major reduction in sizes now. Best we can do is reduce the model size but that's not something this package is made for.

Yes I agree the power is in pytorch themselves, for the problem of assets I will try to split app and models so I download models at run time to show true app size 


Sadly I can't do it now and don't think I will be able for the next days 



Anyway thank you very much for your contribution and for checking the possibility of reducing it to that depth, hopefully pytorch reduces their lib size

**Comment by noob8boi** [comment]
> > > ▒▒▒▒▒▒▒
> > 
> > 
> > As you can see 83MB is taken by the flutter assets which I believe is from the model. So I don't think the package can perform any further optimizations over there. And `armabi-v7a` file is only around 25MB.. so the `libpytorch_jni_lite.so` is definitely not 70MB .. more like 15MB or so. The analysis json file should be able give in depth view with the help of dart dev tools 🤔.. I hope I'm not drawing incorrect conclusions. Basically I don't think there can be major reduction in sizes now. Best we can do is reduce the model size but that's not something this package is made for.
> 
> Yes I agree the power is in pytorch themselves, for the problem of assets I will try to split app and models so I download models at run time to show true app size
> 
> Sadly I can't do it now and don't think I will be able for the next days
> 
> Anyway thank you very much for your contribution and for checking the possibility of reducing it to that depth, hopefully pytorch reduces their lib size

happy to help :)

**Comment by zezo357** [comment]
Btw I will leave your pr open, since it's valid until  the assets stuff is handled if it's better then I guess we have a solution
